apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/bytebuddy/CacheableTransformerDecorator.java

```diff
@@ -0,0 +1,192 @@
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.lang.instrument.IllegalClassFormatException;
+import java.nio.file.Files;
+import java.security.ProtectionDomain;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Wrapper classFileTransformer of ByteBuddy, save the enhanced bytecode to memory cache or file cache,
+ * and automatically load the previously generated bytecode during the second retransform,
+ * to solve the problem that ByteBuddy generates auxiliary classes with different random names every time.
+ * Allow other javaagent to enhance those classes that enhanced by SkyWalking agent.
+ */
+public class CacheableTransformerDecorator implements AgentBuilder.TransformerDecorator {
+
+    private static final ILog logger = LogManager.getLogger(CacheableTransformerDecorator.class);
+
+    private String cacheDirBase;
+    private final ClassCacheMode cacheMode;
+    private ClassCacheResolver cacheResolver;
+
+    public CacheableTransformerDecorator(ClassCacheMode cacheMode) throws IOException {
+        this.cacheMode = cacheMode;
+        initClassCache();
+    }
+
+    public CacheableTransformerDecorator(ClassCacheMode cacheMode, String cacheDirBase) throws IOException {
+        this.cacheDirBase = cacheDirBase;
+        this.cacheMode = cacheMode;
+        initClassCache();
+    }
+
+    private void initClassCache() throws IOException {
+        if (this.cacheMode.equals(ClassCacheMode.FILE)) {
+            File cacheDir;
+            if (this.cacheDirBase == null) {
+                cacheDir = Files.createTempDirectory("class-cache").toFile();
+            } else {
+                cacheDir = new File(this.cacheDirBase + "/class-cache-" + RandomString.make());
+            }
+            cacheResolver = new FileCacheResolver(cacheDir);
+        } else {
+            cacheResolver = new MemoryCacheResolver();
+        }
+    }
+
+    @Override
+    public ResettableClassFileTransformer decorate(ResettableClassFileTransformer classFileTransformer) {
+        return new ResettableClassFileTransformer.WithDelegation(classFileTransformer) {
+
+            @Override
+            public byte[] transform(ClassLoader loader, String className, Class<?> classBeingRedefined, ProtectionDomain protectionDomain, byte[] classfileBuffer) throws IllegalClassFormatException {
+                // load from cache
+                byte[] classCache = cacheResolver.getClassCache(loader, className);
+                if (classCache != null) {
+                    return classCache;
+                }
+
+                //transform class
+                classfileBuffer = classFileTransformer.transform(loader, className, classBeingRedefined, protectionDomain, classfileBuffer);
+
+                // save to cache
+                if (classfileBuffer != null) {
+                    cacheResolver.putClassCache(loader, className, classfileBuffer);
```

Comment:
the `key` should be classloader+className,or maybe duplicate in different classloader